### PR TITLE
Create first_missing_positive.py

### PR DESCRIPTION
**Pull request by moaldeen**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by cclauss** [comment]
```
>>> def x(nums):
...     nums = sorted(num for num in nums if num > 0)
...     return sorted(set(range(1, max(nums) + 2)) - set(nums))[0]
...
>>> x([2, 5, -10, 8, 4, 6, 1, 9])
3
>>> x([7, 8, 9, 11, 12])
1
>>> x([3, 4, -1, 1])
2
>>> x([1, 2, 0])
3
```

File: data_structures/arrays/first_missing_positive.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    Hints:
+    - Try to place each number in its correct position, so that nums[i] equals i + 1.
+    - Iterate through the array to find the first position where nums[i] != i + 1,
+      indicating the missing positive integer.
+
+    :param nums: The input unsorted integer array.
+    :return: The smallest missing positive integer.
+
+    Examples:
+    >>> first_missing_positive([1, 2, 0])
```
Comment (cclauss) [inline data_structures/arrays/first_missing_positive.py:16]: Please add doctests for
```
>>> first_missing_positive([])
>>> first_missing_positive([0])
>>> first_missing_positive([1])
>>> first_missing_positive([1.1, 2.2, 3.3])
>>> first_missing_positive([-1, -2, -3])
```

Comment (moaldeen) [reply data_structures/arrays/first_missing_positive.py:16]: thanks for the feedback .. i have done it

File: data_structures/arrays/first_missing_positive.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
-def first_missing_positive(nums: list[int]) -> int:
```
Comment (cclauss) [inline data_structures/arrays/first_missing_positive.py]: Removing type hints is a bad idea.

**Review by moaldeen — COMMENTED** [review]
hint was added

File: data_structures/arrays/first_missing_positive.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+def first_missing_positive(nums: list[float]) -> int:
+    """
+    Given an unsorted float array nums, returns the smallest missing positive integer.
```
Comment (cclauss) [inline data_structures/arrays/first_missing_positive.py:3]:
```suggestion
    Given an unsorted integer array nums, returns the smallest missing positive integer.
```

File: data_structures/arrays/first_missing_positive.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    The algorithm must run in O(n) time and use O(1) auxiliary space.
+
+    Hints:
+    - Try to place each number in its correct position, so that nums[i] equals i + 1.
+    - Iterate through the array to find the first position where nums[i] != i + 1,
+      indicating the missing positive integer.
+
+    :param nums: The input unsorted float array.
```
Comment (cclauss) [inline data_structures/arrays/first_missing_positive.py:12]:
```suggestion
    :param nums: The input unsorted integer array.
```

File: data_structures/arrays/first_missing_positive.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    >>> first_missing_positive([])
+    1
+
+    >>> first_missing_positive([0])
+    1
+
+    >>> first_missing_positive([1])
+    2
+
+    >>> first_missing_positive([1.5, -2.3, 3.0, 2.7])
```
Comment (cclauss) [inline data_structures/arrays/first_missing_positive.py:34]:
```suggestion
    >>> first_missing_positive([0.9, 1.5, -2.3, 3.0, 2.7])
```

**Review by moaldeen — COMMENTED** [review]
fixed the hint

**Review by moaldeen — COMMENTED** [review]
hint added

**Review by cclauss — CHANGES_REQUESTED** [review]

**Pull request closed without merging**